docs: expand message passing current efforts (#35)
ethereum/L2-interop: docs/message-passing/messaging-current-efforts.md

    MailboxB->>Recipient: handle(origin, sender, _message)
```

## External Messaging Protocols

> Different protocols have been built to enable message passing between the chains they operate on. In principle, all of them function with their own security models and independently of the chain's operations and governance. Most prioritize speed over canonical bridges, as they operate under some degree of confirmation subjectivity.

Reviewer: What is the purpose of this section? There are more protocols than what are listed here (Axelar and wormhole are not mentioned for example). But surely you do not mean to add an exhaustive list of all existing messaging protocols? It's unclear to me what this section is meant to add to the document.

Author: Some protocols (both canonical and external) are listed as examples of existing message interfaces. I'm not trying to include all of them at once, but I'm definitely happy to add anyone who wants to be part of it

sequenceDiagram
    participant User
    participant OriginService as L1 or L2 MessageService (Chain A)
    participant Relayer as PostBots/Off-chain Relayer
    participant DestinationService as L2 or L1 MessageService (Chain B)
    participant Recipient

    User->>OriginService: sendMessage(to, fee, calldata)
    OriginService->>OriginService: emit MessageSent(from, to, fee, value, nonce, calldata, messageHash)
    Note over OriginService: Bridging logic <br/> (L1 deposit or L2 withdrawal initiated)
    OriginService->>Relayer: Off-chain bridging or PostBots flow
    Note over Relayer: Validation flow
    Relayer->>DestinationService: claimMessage(from, to, fee, value, feeRecipient, calldata, nonce) <br/> or claimMessageWithProof(ClaimMessageWithProofParams params) 
    DestinationService->>DestinationService: emit MessageClaimed(messageHash)
    DestinationService->>Recipient: Deliver message contents
    DestinationService->>Relayer: receive fees (if it is allowed/seted)

```

### OP Stack

OP Stack deploys corresponding messenger contracts on both L1 and L2, as well as L2-to-L2 when interoperability is enabled at the protocol level. All of them include replay protection.

> - **[L1→L2 / L2→L1](https://specs.optimism.io/protocol/messengers.html)**: This follows the `CrossDomainMessenger` library. The `sendMessage` function includes the target, value (in ETH), message data, and gas limit. The `relayMessage` function, in turn, includes the same parameters while adding the sender and nocce.

Reviewer 2: ```suggestion
- **[L1→L2 / L2→L1](https://specs.optimism.io/protocol/messengers.html)**: This follows the `CrossDomainMessenger` library. The `sendMessage` function includes the target, value (in ETH), message data, and gas limit. The `relayMessage` function, in turn, includes the same parameters while adding the sender and nonce.
```

    - L1→L2 operates on a push-based model, where sequencers process deposits when deemed safe, strictly following the order in which they were initiated.

>     - L2→L1 follows a pull-based model, where withdrawals are in practice finalized asynchronously, executed through `finalizeWithdrawalTransaction`.

Reviewer 2: ```suggestion
    - L2→L1 follows a pull-based model, where withdrawals are finalized asynchronously in practice, and executed through `finalizeWithdrawalTransaction`.
```

    OriginMessenger->>OriginMessenger: emit SentMessage(Sender, target, value, nonce, gasLimit, message)
    Note over OriginMessenger: Bridging logic <br/> (L1 deposit or L2 withdrawal initiated)
    OriginMessenger->>Relayer: Off-chain proof of inclusion/finality reached
    Note over Relayer: Validation flow
    Relayer->>DestinationMessenger: relayMessage(from, to, value, nonce, message) <br/> or relayMessageWithProof(..., proof)
    DestinationMessenger->>DestinationMessenger: emit RelayedMessage(messageHash)
    DestinationMessenger->>Recipient: Execute message contents

```

# Comparison

The existing approaches are divided into proposed standards and actual implementations. While standards aim to be agnostic and broad to cover a wide range of use cases, actual implementations serve as references for what is included and how it is implemented.

| **Standard (or protocol)** | **ERC-6170** | **ERC-7786** | **ERC-7841** | **ERC-7854** | **CCTP V1** | **CCTP V2** | **LayerZero** | **OP Stack L1/L2** | **OP Stack L2/L2** | **Linea** | **Scroll** |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |

> | **What is it? (*Status*)** | Standard (Draft, merged at Dec 25, 2022) | Standard (Draft, merged at Dec 4, 2024) | Standard (Draft, open) | Standard (Draft, open) | Protocol (Implemented) | Protocol (Implemented) | Protocol (Implemented) | Protocol (Implemented) | Protocol (Devnet) | Protocol (Implemented) | Protocol (Implemented) |

Reviewer 2: ```suggestion
| **What is it? (*Status*)** | Standard (Draft, merged on Dec 25, 2022) | Standard (Draft, merged on Dec 4, 2024) | Standard (Draft, open) | Standard (Draft, open) | Protocol (Implemented) | Protocol (Implemented) | Protocol (Implemented) | Protocol (Implemented) | Protocol (Devnet) | Protocol (Implemented) | Protocol (Implemented) |
```

> | **Motivation / Main Use Case** | Interface for message passers. | Interface for message passers. | Interface for message passers. Focused for L2s. | Interface for message passers. Focused on Ethereum ecosystem. | Message Layer for USDC cross-chain transfers. | Message Layer for USDC cross-chain transfers. | Message Layer for any cross-chain use cases. | Message Layer leveraged from underlying rollup security model. | Message Layer leveraged from underlying rollup security model. | Message Layer leveraged from underlying rollup security model. | Message Layer leveraged from underlying rollup security model. |

Reviewer 2: ```suggestion
| **Motivation / Main Use Case** | Interface for message-passing | Interface for message-passing | Interface for Interface for message-passing, focused for L2s. | Interface for message-passing, focused on Ethereum ecosystem. | Message Layer for USDC cross-chain transfers. | Message Layer for USDC cross-chain transfers. | Message Layer for any cross-chain use cases. | Message Layer leveraged from underlying rollup security model. | Message Layer leveraged from underlying rollup security model. | Message Layer leveraged from underlying rollup security model. | Message Layer leveraged from underlying rollup security model. |
```

> | **VM Compatibility** | Intended to be agnostic. Usage of `bytes`  | Intended to be agnostic. It uses CAIP for this purpose. | Intended to be agnostic. It cointain a rationale explaining how to make it work in other VMs different from EVM. | Intended to be agnostic. ISM is modular to couple. | Multiple VMs. | Multiple VMs. | Multiple VMs. | - | - | - | - |

Reviewer 2: ```suggestion
| **VM Compatibility** | Intended to be agnostic. Usage of `bytes`  | Intended to be agnostic. It uses CAIP for this purpose. | It includes a rationale on how to adapt it to virtual machines other than the EVM.| Intended to be agnostic. ISMs are modular and extensible | Multiple VMs. | Multiple VMs. | Multiple VMs. | - | - | - | - |
```

| **Gas/Fee Handling** | None built-in. Payable function + `data_` field can hold fee info. | Payable. `value` in event used for fees, plus flexible `attributes`. There is a value field. | None built-in. | Fee mechanism via its post-dispatch hooks (e.g. through `INTERCHAIN_GAS_PAYMASTER`) and add a quote function. | No fees. | No direct fees but hooks can include a payment logic. | Fee pay out in origin in `send`. | Deposits include L2 gas. Withdrawals cost are covered by who call `relayMessage`. | Not defined. | User may leave a tip, optional. | Deposits include L2 gas. Withdrawals cost are covered by who call `relayMessage`. |
| **Bundling** | Single message. | Single message, but attributes can define some bundling logic. | `sessionId` can group messages. No atomic multi-call. | Yes, possible bundling through ISM validation to orchestrate it. | No bundling. | Yes, limited to “mint + 1 contract call”. | Single message, but attributes can define some bundling logic. `guid`/`compose`can coordinate multiple packets. | Single call invocation. | Single call invocation. | Single call invocation. | Single call invocation. |

> | **Pull/Push Support** | Not defined (but it could given its simplicity) | Explicitly supports both. | Supported via separate mailbox contracts | Supported; e.g. integration with synchronous L2s like Superchain outlined. | Support both. | Support both. | Support both. | Push in deposits. Pull in withdrawals. | Support both. | Push in deposits. Pull still available. Pull in withdrawals. | Push in deposits. Pull in withdrawals. |

Reviewer 2: ```suggestion
| **Pull/Push Support** | Not defined (but it could given its generality) | Explicitly supports both. | Supported via separate mailbox contracts | Supported; e.g. integration with synchronous L2s like Superchain outlined. | Support both. | Support both. | Support both. | Push in deposits. Pull in withdrawals. | Support both. | Push in deposits. Pull still available. Pull in withdrawals. | Push in deposits. Pull in withdrawals. |
```
